code/game/objects/items/airbubble.dm

```diff
@@ -0,0 +1,492 @@
+
+	for(var/mob/M in src)
+		M.forceMove(loc)
+		if(M.client)
+			M.client.eye = M.client.mob
+			M.client.perspective = MOB_PERSPECTIVE
+
+/obj/structure/closet/airbubble/Initialize()
+	. = ..()
+	add_inside()
+	START_PROCESSING(SSfast_process, src)
+
+/obj/structure/closet/airbubble/Destroy()
+	STOP_PROCESSING(SSfast_process, src)
+	qdel(internal_tank)
+	if(parts)
+		new parts(loc)
+	if (smooth)
+		queue_smooth_neighbors(src)
+	return ..()
+
+/obj/structure/closet/airbubble/close()
+	if(!opened)
+		return 0
+	if(!can_close())
+		return 0
+
+	var/stored_units = 0
+
+	if(store_misc)
+		stored_units += store_misc(stored_units)
+	if(store_items)
+		stored_units += store_items(stored_units)
+	if(store_mobs)
+		stored_units += store_mobs(stored_units)
+
+	icon_state = icon_closed
+	opened = 0
+
+	playsound(loc, close_sound, 25, 0, -3)
+	density = 1
+	add_inside()
+	return 1
+
+// Fold the bubble, transfering properties.
+/obj/structure/closet/airbubble/MouseDrop(over_object, src_location, over_location)
+	if((!zipped || ripped )&& (over_object == usr && (in_range(src, usr) || usr.contents.Find(src))))
+		if(!ishuman(usr))	return
+		if(opened)	return 0
+		if(contents.len > 1)	return 0
+		visible_message("[usr] folds up the [src.name]")
+		var/obj/item/airbubble/bag = new /obj/item/airbubble(get_turf(src))
+		bag.ripped = ripped
+		bag.zipped = zipped
+		bag.internal_tank = internal_tank
+		internal_tank.forceMove(bag)
+		internal_tank = null
+		bag.w_class = ITEMSIZE_LARGE
+
+		if(ripped)
+			bag.icon_state = "[icon_closed]_man_folded_ripped"
+			bag.desc = "Special air bubble designed to protect people inside of it from decompressed enviroments. It appears to be poorly hand folded. <span class='danger'>It has hole in it! Maybe you shouldn't use it!</span>"
+		else
+			bag.icon_state = "[icon_closed]_man_folded"
+			bag.desc = "Special air bubble designed to protect people inside of it from decompressed enviroments. It appears to be poorly hand folded."
+		qdel(src)
+		return
+
+/obj/structure/closet/airbubble/req_breakout()
+
+	if(opened)
+		return 0 //Door's open... wait, why are you in it's contents then?
+	if(zipped)
+		return 1 // If it is zipped
+	if(breakout)
+		return -1 //Already breaking out.
+	return 0
+
+// If we are stuck, and need to get out
+/obj/structure/closet/airbubble/mob_breakout(var/mob/living/escapee)
+
+	breakout_time *= req_breakout()
+	if(breakout_time <= 0)
+		return
+
+	escapee.next_move = world.time + 100
+	escapee.last_special = world.time + 100
+	escapee << "<span class='warning'>You lean on the back of \the [src] and start punching internal wall with your legs. (this will take about [breakout_time] minutes)</span>"
+	visible_message("<span class='danger'>\The [src] begins to shake violently! Something is terring it from the inside!</span>")
+
+	var/time = 6 * breakout_time * 2
+
+	var/datum/progressbar/bar
+	if (escapee.client && escapee.client.prefs.toggles_secondary & PROGRESS_BARS)
+		bar = new(escapee, time, src)
```

Comment:
Is this level of reinvention necessary? `do_after` has an argument `extra_checks`. It can be used to do what's required. It eager-returns if they fail. So basically:

```dm
/obj/structure/closet/airbubble/proc/breakout_callback(/mob/living/escapee)
	if (QDELETED(escapee))
		return FALSE

	if ((world.time - last_shake) > 5 SECONDS)
		playsound(loc, "sound/items/[pick("rip1","rip2")].ogg", 100, 1)
		animate_shake()
		last_shake = world.time

	if (!req_breakout())
		return FALSE

	return TRUE

/obj/structure/closet/airbubble/mob_breakout(/mob/living/escapee)
	if (req_breakout() < 1)
		return

	escapee.next_move = world.time + 100
	escapee.last_special = world.time + 100
	escapee << "<span class='warning'>You lean on the back of \the [src] and start punching internal wall with your legs. (this will take about [breakout_time] minutes)</span>"
	visible_message("<span class='danger'>\The [src] begins to shake violently! Something is terring it from the inside!</span>")

	var/time = 6 * breakout_time * 2
	breakout = TRUE

	if (!do_after(escapee, time, extra_checks = CALLBACK(src, .proc/breakout_callback, escapee)))
		breakout = FALSE
		return

	breakout = FALSE
	escapee << "<span class='warning'>You successfully break out! Tearing the bubble's walls!</span>"
	visible_message("<span class='danger'>\the [escapee] successfully broke out of \the [src]! Tearing the bubble's walls!</span>")
	playsound(loc, "sound/items/[pick("rip1","rip2")].ogg", 100, 1)
	break_open()
	animate_shake()
	qdel(bar)
```